specifications/FiniteMonotonic/Finitize_CRDT.tla

```diff
@@ -0,0 +1,72 @@
+--------------------------- MODULE Finitize_CRDT ----------------------------
+
+EXTENDS Naturals
+
+CONSTANTS Node, Divergence
+
+VARIABLES counter, converge
+
+vars == <<counter, converge>>
+
+S == INSTANCE CRDT
+
+TypeOK ==
+  /\ S!TypeOK
+  /\ counter \in [Node -> [Node -> 0 .. Divergence]]
+  /\ converge \in BOOLEAN
+
+Safety == S!Safety
+
+Monotonicity == [][
+  \/ S!Monotonic
+  \/ \A a, b, c, d \in Node :
+    (counter'[a][b] - counter[a][b]) = (counter'[c][d] - counter[c][d])
+]_vars
+
+Liveness == converge ~> S!Convergence
+
+Init ==
+  /\ S!Init
+  /\ converge = FALSE
+
+Increment(n) ==
+  /\ ~converge
+  /\ counter[n][n] < Divergence
```

Comment:
This is effectively the state constraint in Constraint_CRDT in addition to being one of Increment's enablement conditions.

Reply:
Yes, this keeps the counters from going beyond the divergence limit and thus ensuring the state space remains finite

Reply:
I don't understand why this is less of a hammer than what the blog post calls a "state restriction hammer".

By the way, we also "hardcode" the state constraints in [this model](https://github.com/microsoft/CCF/blob/999ce8fb755e64e977f6224398d1f87fdf5348de/tla/consensus/MCccfraft.tla#L47-L107), for example, to prevent TLC from evaluating an action when the original state constraint would evaluate to false. While this approach improves efficiency slightly, the impact is minimal.  It does lower the value of "states generated", though.

Reply:
So you say that adding a state constraint along with a GarbageCollect action would be equivalent to having the line `counter[n][n] < Divergence` plus a GarbageCollect action?

Reply:
Yes, TLC consistently reports the same outcome (counterexample/pass), when exploring the different fairness constraints in CRDT below:

```tla
------------------------------- MODULE CRDT ---------------------------------
EXTENDS Naturals

CONSTANT Node

VARIABLE counter
vars == counter

TypeOK == counter \in [Node -> [Node -> Nat]]

Safety == \A n, o \in Node : counter[n][n] >= counter[o][n]

Convergence == []<>(\A n, o \in Node : counter[n] = counter[o])

Init == counter = [n \in Node |-> [o \in Node |-> 0]]

Increment(n) == counter' = [counter EXCEPT ![n][n] = @ + 1]

Gossip(n, o) ==
  LET Max(a, b) == IF a > b THEN a ELSE b IN
  counter' = [
    counter EXCEPT ![o] = [
      nodeView \in Node |->
        Max(counter[n][nodeView], counter[o][nodeView])
      ]
    ]

Next ==
  \/ \E n \in Node : Increment(n)
  \/ \E n, o \in Node : Gossip(n, o)

Fairness ==
    \* /\ WF_vars(Next)
    \* /\ ~ \E n \in Node : <>[][Increment(n)]_vars
    \* /\ \A n, o \in Node : <>[][Gossip(n,o)]_vars
    \* /\ \A n \in Node : WF_vars(Increment(n))
    /\ \A n, o \in Node : WF_vars(Gossip(n,o))
    \* /\ Convergence

Spec ==
  /\ Init
  /\ [][Next]_counter
  /\ Fairness

=============================================================================
------ MODULE MCCRDT -----
EXTENDS CRDT

CONSTANT Divergence

C == INSTANCE CRDT

MCIncrement(n) ==
  /\ counter[n][n] < Divergence
  /\ C!Increment(n)

GarbageCollect ==
  LET SetMin(s) == CHOOSE e \in s : \A o \in s : e <= o IN
  LET Transpose == SetMin({counter[n][o] : n, o \in Node}) IN
  /\ counter' = [
      n \in Node |-> [
        o \in Node |-> counter[n][o] - Transpose
      ]
    ]

MCNext ==
  \/ C!Next
  \/ GarbageCollect

StateConstraint == \A n, o \in Node : counter[n][o] <= Divergence

======
```
```
SPECIFICATION Spec

CONSTANTS
  Node = {n1, n2}
  Next <- MCNext

INVARIANTS
    TypeOK
    Safety

PROPERTIES 
    Convergence

CONSTANTS Divergence = 2
CONSTANTS Increment <- MCIncrement

-----

SPECIFICATION Spec

CONSTANTS
  Node = {n1, n2}
  Next <- MCNext

INVARIANTS
    TypeOK
    Safety

PROPERTIES 
    Convergence

CONSTANTS Divergence = 2
CONSTRAINTS StateConstraint
```


*Without* `GarbageCollect`, TLC reports pass with the state constraint but a counterexample with the increment conjunct with the `\A n \in Node : WF_vars(Increment(n))` fairness constraint.